File: go/stack/856_score_of_parentheses.go
package stack

// scoreOfParentheses 括号的分数
// Method1: 栈
func scoreOfParentheses(s string) int {

	// 括号的分数，相当于把把括号看成树，每个节点的值，等于子节点之和* 2
	// 求虚拟根节点的值

	ans := []int{0}

	for _, ch := range s {
		// 左括号压栈，并赋值初试分数为0
		if ch == '(' {
			ans = append(ans, 0)
		} else {
			// 碰到右括号(有匹配项) 出栈，出栈处的对应的左( 计算分数
			// 栈顶剩下左括号的分数 = max(2 * 出栈的(左括号内部的分数)， 1)
			
			v := ans[len(ans)-1]
			ans = ans[:len(ans)-1]
			// 所有子节点分数之和
			ans[len(ans)-1] += max(2*v, 1)
		}
	}

	return ans[0]
}

func max(x, y int) int {
	if x > y {
		return x
	}

	return y
}

// scoreOfParentheses2 根节点的值是 每个叶子节点值 与其深度的幂的总和。
func scoreOfParentheses2(s string) int {

	var ans, depth int

	for i, ch := range s {
		if ch == '(' {
			depth += 1
		} else {
			depth--

			if ch == ')' && s[i-1] == '(' {
				ans += 1 << depth
			}
		}
	}

	return ans
}

File: rust/src/stack/856_score_of_parentheses.rs


// score_of_parentheses 栈的用法，碰到右括号，计算这对括号对
// 上对括号的贡献值
pub fn score_of_parentheses(s: String) -> i32 {

    let mut ans = vec![0];

    for ch in s.chars() {
        if ch == '(' {
            ans.push(0);
        }else {
            // 先删除对应的左括号, 然后计算内部的分数
            
            if let Some(v)= ans.pop() {
                // 更新 这对删除的 平衡左括对ans 的分数影响 +=
                if let Some(last) = ans.last_mut() {
                    // *last += std::cmp::max(2 * v, 1);
                    *last += 1.max(2 * v)
                }
            }
        }
    }

    ans[0]
}

// 当作树 => 根节点的值是 每个叶子节点值 与其深度的幂的总和。
pub fn score_of_parentheses2(s: String) -> i32 {

    let mut ans = 0;
    let mut depth = 0;

    let chars: Vec<char> = s.chars().collect();

    for (i, &ch) in chars.iter().enumerate(){
        if ch == '(' {
            depth += 1;
        }else {
            depth -= 1;
            
            // 这对平衡括号的值，等于2 的其深度的幂次方
            if ch == ')' && chars[i-1] == '(' {
                ans += 1 << depth;
            }
        }
    }

    ans
}